R/truncate_linelist.R

```diff
@@ -59,71 +43,84 @@
     )
-  }
-  if (!inherits(max_date, "Date")) {
-    # enables numeric max_date for R <4.3.0
-    # https://github.com/wch/r-source/commit/920affdc32cba1baa765fb3a647570d407dfccaa # nolint line_length_linter
-    max_date <- as.Date(max_date, origin = "1970-01-01")
-    message(
-      "Truncation max date is: ", max_date, ".\n",
-      "Assuming an origin of '1970-01-01' in line with R >= v4.3.0."
+  unit <- match.arg(unit)
+  direction <- match.arg(direction)
+
+  if (is.numeric(truncation_time)) {
+    # convert truncation_time to days
+    truncation_time <- switch(unit,
+      days = truncation_time,
+      weeks = truncation_time * 7,
+      months = truncation_time * (365.25 / 12),
+      years = truncation_time * 365.25
     )
+
+    date_cols <- grep(pattern = "date_", x = colnames(linelist), fixed = TRUE)
+    if (direction == "backwards") {
+      # get present as maximum date in line list
+      max_date <- as.Date(
+        max(unlist(linelist[, date_cols]), na.rm = TRUE),
+        origin = "1970-01-01"
+      )
+      trunc_date <- max_date - truncation_time
+    } else {
+      # get outbreak start date as minimum date in line list
+      min_date <- as.Date(
+        min(unlist(linelist[, date_cols]), na.rm = TRUE),
+        origin = "1970-01-01"
+      )
+      trunc_date <- min_date + truncation_time
+    }
+  } else {
+    trunc_date <- truncation_time
+    if (!arg_ignore) {
+      warning(
+        "When `truncation_time` is given as a <Date>, ",
+        "`unit` and `direction` are ignored.",
+        call. = FALSE
+      )
+    }
   }
-  col_name <- paste0("date_", truncation_event)
-  trunc_time <- delay(nrow(linelist))
-  # sample which onset dates are longer than reporting delay (i.e. reported)
-  reported_lgl_idx <-
-    (max_date - linelist[[col_name]]) > trunc_time
-  # convert NAs to TRUE to prevent issues with subsetting with NAs
-  reported_lgl_idx[is.na(reported_lgl_idx)] <- TRUE
-  linelist <- linelist[reported_lgl_idx, ]
 
-  # subset truncation times to remove times for individuals removed above
-  trunc_time <- trunc_time[reported_lgl_idx]
+  # which cases are reported before the truncation date (i.e. included)
+  reported_lgl_idx <- trunc_date > linelist$date_reporting
+  linelist <- linelist[reported_lgl_idx, ]
 
   # convert events (reporting, admissions & outcomes) more recent than
   # truncation time to NA
-  missing_outcome_lgl_idx <- (max_date - linelist$date_outcome) < trunc_time
-  linelist$date_outcome[missing_outcome_lgl_idx] <- NA_real_
-  missing_admission_lgl_idx <- (max_date - linelist$date_admission) < trunc_time
-  linelist$date_admission[missing_admission_lgl_idx] <- NA_real_
-  missing_reporting_lgl_idx <- (max_date - linelist$date_reporting) < trunc_time
-  linelist$date_reporting[missing_reporting_lgl_idx] <- NA_real_
+  linelist$date_outcome[linelist$date_outcome > trunc_date] <- NA_real_
+  linelist$date_admission[linelist$date_admission > trunc_date] <- NA_real_
```

Comment:
could make this a loop over fields starting with `date_` for future proofing?

Reply:
Resolved in 1a202ccaba1728e9d119364e6bb790cf0d824eea.